### ISSUE TITLE
[libafl_qemu] Emulator::mappings returns incorrect / incomplete results with qemu-i386

### ISSUE THREAD

**Issue by RickdeJager** [AFLplusplus/LibAFL#575]
**Version**
I am using the latest version of the main branch at the time of writing. (specifically c72f773ca02c4038aed0691dc5e89f643fb11e38)

**Describe the bug**
When using libafl_qemu with `i386`, `mappings()` is not populated correctly. Specifically, calling `end()` on a `MapInfo` object will always return 0. The bug does not appear when using libafl_qemu with `x86_64`.

**To Reproduce**
I created a small test program to reproduce the issue:
```rust
// Cargo.toml: libafl_qemu = { path = "../LibAFL/libafl_qemu/", features = ["i386"] }

use libafl::Error;
use libafl_qemu::Emulator;

fn main() -> Result<(), Error> {
    let args = vec![
        "qemu-i386".to_string(),
        // Minimal test binary, consists of just a single line `int main() {}`.
        // compile command: `gcc -m32 test.c -o test_32 -no-pie`
        "test_32".to_string(),
    ];

    let env: Vec<(String, String)> = Vec::new();
    let emu = Emulator::new(&args, &env);

    for map in emu.mappings() {
        println!("Mapping: {:x} -> {:x}", map.start(), map.end());
    }
    Ok(())
}
```

Running the program with `cargo run` yields the following output:
```
Mapping: 8058000 -> 0
Mapping: 805b000 -> 0
Mapping: 3ffe0000 -> 0
Mapping: 3ffe2000 -> 0
Mapping: 3ffe3000 -> 0
Mapping: 4000d000 -> 0
Mapping: 4000e000 -> 0
Mapping: 40010000 -> 0
Mapping: 40011000 -> 0
```

**Expected behavior**
`map.end()` should return a non-zero value. For instance, here is the output I receive if I convert the above code snippet to use `qemu-x86_64` instead:
```
Mapping: 403000 -> 405000
Mapping: 4000000000 -> 4000001000
Mapping: 4000001000 -> 4000801000
Mapping: 4000801000 -> 400082d000
Mapping: 400082d000 -> 400082e000
Mapping: 400082e000 -> 4000830000
Mapping: 4000830000 -> 4000833000
```

**Additional context**
This bug also breaks the `QemuSnapshotHelper` on `i386` since the snapshot function contains the following line of code:
```rust
// libafl_qemu/src/snapshot.rs:73
// --- < snip > ---
        for map in emulator.mappings() {
            let mut addr = map.start();
            while addr < map.end() {    // always false
// --- < snip > ---
```

**Comment by RickdeJager** [comment]
I looked into this a bit further and managed to come up with a partial fix (and an additional bug I initially didn't notice :upside_down_face:).

**root cause**
Firstly, the bug occurs because the `MapInfo`structs size is depended on the size of `GuestAddr`. In a 32 bit config, this structs members no longer match their C representation (which uses all u64's). So when libafl-qemu goes to populate the struct, it writes too much data, causing both alignment issues and probably some corruption as well.

**Rust struct**
```rust
#[repr(C)]
#[cfg_attr(feature = "python", pyclass(unsendable))]
pub struct MapInfo {
    start: GuestAddr,     // = u32
    end: GuestAddr,      // = u32
    offset: GuestAddr,   // = u32
    path: *const u8,
    flags: i32,
    is_priv: i32,
}
```
** c struct **
```c
struct libafl_mapinfo {
    uint64_t start, end;
    uint64_t offset;
    const char* path;
    int flags, is_priv;
};
```

** Temporary fix (so far)**
I changed the `GuestMaps` iterator to use a temp struct that exactly matches the qemu struct, and manually cast each member to GuestAddrs. 
```rust
#[repr(C)]
pub struct TmpMapInfo {
    start: u64,
    end: u64,
    offset: u64,
    path: *const u8,
    flags: i32,
    is_priv: i32,
}

impl Iterator for GuestMaps {
    type Item = MapInfo;

    #[allow(clippy::uninit_assumed_init)]
    fn next(&mut self) -> Option<Self::Item> {
        if self.c_iter.is_null() {
            return None;
        }
        unsafe {
            let mut tmp: TmpMapInfo = MaybeUninit::uninit().assume_init();
            self.c_iter = libafl_maps_next(self.c_iter, addr_of_mut!(tmp));
            if self.c_iter.is_null() {
                None
            } else {
                Some(MapInfo {
                    // I am not sure why this 0x10000 is needed yet, see below
                    start: (tmp.start - 0x10000) as GuestAddr,
                    end: (tmp.end - 0x10000) as GuestAddr,
                    offset: tmp.offset as GuestAddr,
                    path: tmp.path,
                    flags: tmp.flags,
                    is_priv: tmp.is_priv,
                })
            }
        }
    }
}
```

**Todo**
This patch seems to work fine, but I also found that the addresses were always off by exactly `0x10000`. To prevent segmentation faults in the snapshot call, I've added the subtraction, which _seems_ to work fine, but I have not figured out where or why this offset is added.

**Comment by andreafioraldi** [comment]
check if 0x10000 is emu::load_addr(), if so you ave to use h2g

**Comment by RickdeJager** [comment]
Ah that makes sense. As a quick test, I manually subtracted `guest_base` from the `start` and `end` addresses, which seems to work fine.  
The `Guestmaps` struct does not have a reference to the Emulator, so I don't think I can easily call `h2g` in this context.

**Comment by andreafioraldi** [comment]
fix in #619
